pkg/controller/persistentvolumelabel/pvlcontroller.go

```diff
@@ -0,0 +1,382 @@
+
+	// start up your worker threads based on threadiness.  Some controllers have multiple kinds of workers
+	for i := 0; i < threadiness; i++ {
+		// runWorker will loop until "something bad" happens.  The .Until will then rekick the worker
+		// after one second
+		go wait.Until(pvlc.runWorker, time.Second, stopCh)
+	}
+
+	// wait until we're told to stop
+	<-stopCh
+}
+
+func (pvlc *PersistentVolumeLabelController) runWorker() {
+	// hot loop until we're told to stop.  processNextWorkItem will automatically wait until there's work
+	// available, so we don't worry about secondary waits
+	for pvlc.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (pvlc *PersistentVolumeLabelController) processNextWorkItem() bool {
+	// pull the next work item from queue.  It should be a key we use to lookup something in a cache
+	key, quit := pvlc.queue.Get()
+	if quit {
+		return false
+	}
+	// you always have to indicate to the queue that you've completed a piece of work
+	defer pvlc.queue.Done(key)
+
+	// do your work on the key.  This method will contains your "do stuff" logic
+	err := pvlc.syncHandler(key.(*v1.PersistentVolume))
+	if err == nil {
+		// if you had no error, tell the queue to stop tracking history for your key.  This will
+		// reset things like failure counts for per-item rate limiting
+		pvlc.queue.Forget(key)
+		return true
+	}
+
+	// there was a failure so be sure to report it.  This method allows for pluggable error handling
+	// which can be used for things like cluster-monitoring
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
+
+	// since we failed, we should requeue the item to work on later.  This method will add a backoff
+	// to avoid hotlooping on particular items (they're probably still not going to work right away)
+	// and overall controller protection (everything I've done is broken, this controller needs to
+	// calm down or it can starve other useful work) cases.
+	pvlc.queue.AddRateLimited(key)
+
+	return true
+}
+
+// AddLabels adds appropriate labels to persistent volumes and sets the
+// volume as available if successful.
+func (pvlc *PersistentVolumeLabelController) AddLabels(volume *v1.PersistentVolume) error {
+	var volumeLabels map[string]string
+
+	// Only add labels if in the list of initializers
+	if pvlc.shouldLabel(volume) {
+		if volume.Spec.AWSElasticBlockStore != nil {
+			labels, err := pvlc.findAWSEBSLabels(volume)
+			if err != nil {
+				return fmt.Errorf("error querying AWS EBS volume %s: %v", volume.Spec.AWSElasticBlockStore.VolumeID, err)
+			} else {
+				volumeLabels = labels
+			}
+		}
+		if volume.Spec.GCEPersistentDisk != nil {
+			labels, err := pvlc.findGCEPDLabels(volume)
+			if err != nil {
+				return fmt.Errorf("error querying GCE PD volume %s: %v", volume.Spec.GCEPersistentDisk.PDName, err)
+			} else {
+				volumeLabels = labels
+			}
+		}
+
+		if len(volumeLabels) != 0 {
```

Comment:
It looks like we still have the same problem where if no labels are applied with AWS/GCE, or for all non AWS/GCE volumes, we never remove the initializer?